File: common/src/main/java/maestro/api/event/events/TickEvent.java
package maestro.api.event.events;

import java.util.function.BiFunction;
import maestro.api.event.events.type.EventState;
import net.minecraft.client.Minecraft;

/**
 * Called on and after each game tick of the primary {@link Minecraft} instance and dispatched to
 * all Maestro instances.
 *
 * <p>When {@link #state} is {@link EventState#PRE}, the event is being called just prior to when
 * the current in-game screen is ticked. When {@link #state} is {@link EventState#POST}, the event
 * is being called at the very end of the {@link Minecraft#runTick(boolean)} method.
 */
public record TickEvent(EventState state, Type type, int count) {

    private static int overallTickCount;

    public static synchronized BiFunction<EventState, Type, TickEvent> createNextProvider() {
        final int count = overallTickCount;
        overallTickCount++;
        return (state, type) -> new TickEvent(state, type, count);
    }

    public enum Type {
        /** When guarantees can be made about the game state and in-game variables. */
        IN,
        /**
         * No guarantees can be made about the game state. This probably means we are at the main
         * menu.
         */
        OUT,
    }
}

File: common/src/main/java/maestro/api/event/events/TickEvent.kt
package maestro.api.event.events

import maestro.api.event.events.type.EventState

/**
 * Called on and after each game tick of the primary Minecraft instance and dispatched to
 * all Maestro instances.
 *
 * When [state] is [EventState.PRE], the event is being called just prior to when
 * the current in-game screen is ticked. When [state] is [EventState.POST], the event
 * is being called at the very end of the Minecraft.runTick method.
 */
data class TickEvent(
    @JvmField val state: EventState,
    @JvmField val type: Type,
    @JvmField val count: Int,
) {
    enum class Type {
        /** When guarantees can be made about the game state and in-game variables. */
        IN,

        /** No guarantees can be made about the game state. This probably means we are at the main menu. */
        OUT,
    }

    companion object {
        private var overallTickCount = 0

        @JvmStatic
        @Synchronized
        fun createNextProvider(): java.util.function.BiFunction<EventState, Type, TickEvent> {
            val count = overallTickCount++
            return java.util.function.BiFunction { state, type -> TickEvent(state, type, count) }
        }
    }
}
